pkg/controller/elasticsearch/version/zen2/initial_master_nodes.go

```diff
@@ -5,32 +5,182 @@
 package zen2
 
 import (
+	"context"
+	"fmt"
+	"strings"
+
 	esv1 "github.com/elastic/cloud-on-k8s/pkg/apis/elasticsearch/v1"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/common/version"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/bootstrap"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/client"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/label"
 	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/nodespec"
+	"github.com/elastic/cloud-on-k8s/pkg/controller/elasticsearch/sset"
+	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
+)
+
+const (
+	// InitialMasterNodesAnnotation is applied on the Elasticsearch resource while a cluster is
+	// bootstrapping zen2, and removed when bootstrapping is done.
+	initialMasterNodesAnnotation = "elasticsearch.k8s.elastic.co/initial-master-nodes"
 )
 
-// SetupInitialMasterNodes modifies the ES config of the given resources to setup
-// cluster initial master nodes.
-func SetupInitialMasterNodes(
-	nodeSpecResources nodespec.ResourcesList,
-) error {
-	masters := nodeSpecResources.MasterNodesNames()
-	if len(masters) == 0 {
+// SetupInitialMasterNodes sets the `cluster.initial_master_nodes` configuration setting on
+// zen2-compatible master nodes from nodeSpecResources if necessary.
+// This is only necessary when bootstrapping a new zen2 cluster, or when upgrading a single zen1 master.
+// Rolling upgrades from eg. v6 to v7 do not need that setting.
+// It ensures `cluster.initial_master_nodes` does not vary over time, when this function gets called multiple times.
+func SetupInitialMasterNodes(es esv1.Elasticsearch, k8sClient k8s.Client, nodeSpecResources nodespec.ResourcesList) error {
+	if v, err := version.Parse(es.Spec.Version); err != nil || !versionCompatibleWithZen2(*v) {
+		// we only care about zen2-compatible clusters here
+		return err
+	}
+
+	// if the cluster is annotated with `cluster.initial_master_nodes` (zen2 bootstrap in progress),
+	// make sure we reuse that value since it is not supposed to vary over time
+	if initialMasterNodes := getInitialMasterNodesAnnotation(es); initialMasterNodes != nil {
+		return patchInitialMasterNodesConfig(nodeSpecResources, initialMasterNodes)
+	}
+
+	// we want to set `cluster.initial_master_nodes` if either:
+	// - a new cluster is getting created
+	isInitialClusterBootstrap := !bootstrap.AnnotatedForBootstrap(es)
```

Comment:
nit: we later use `!isInitialClusterBootstrap`, there is a double negation which make the code a little bit tricky to understand at first :)

Reply:
Agreed 👍 
I'll change to the opposite: `isAlreadyBootstrapped`